### PR TITLE
[core] Remove connect at start

### PR DESCRIPTION
**Pull request by dentiny**
Preparation for https://github.com/ray-project/ray/issues/47994, which simplifies the state machine via combining two states at initialization.
This PR is the first step, which removes the useless `connect_at_start` option.

### CODE REVIEW

**Comment by rynewang** [comment]
The feature was originally introduced in https://github.com/ray-project/ray/pull/14722 to deflake  python/ray/tests/test_object_spilling.py.

We need to take a look at the original intention and if this PR makes the test flaky.

**Comment by rynewang** [comment]
Will merge for now. If the test_object_spilling tests start to flake, we will revert this.

**Review by rynewang — APPROVED** [review]

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -750,9 +750,8 @@ CoreWorker::CoreWorker(CoreWorkerOptions options, const WorkerID &worker_id)
   // Tell the raylet the port that we are listening on.
   // NOTE: This also marks the worker as available in Raylet. We do this at the
   // very end in case there is a problem during construction.
-  if (options.connect_on_start) {
-    ConnectToRayletInternal();
-  }
+  ConnectToRayletInternal();
```
Comment (jjyao) [inline src/ray/core_worker/core_worker.cc:753]: We should move this to the very end of the constructor.

Comment (dentiny) [reply src/ray/core_worker/core_worker.cc:753]: > This will match the current behavior.

I didn't change any code position..

Comment (rynewang) [reply src/ray/core_worker/core_worker.cc:753]: today's behavior: if driver, connect here. if worker, connect right after ctor.

Comment (jjyao) [reply src/ray/core_worker/core_worker.cc:753]: My understanding is that this PR unifies driver and worker connect and I think following the worker connect position is better: i.e. connect the very end of the constructor.

Comment (dentiny) [reply src/ray/core_worker/core_worker.cc]: I moved to the end of the ctor.

**Review by jjyao — APPROVED** [review]

**Comment by dentiny** [comment]
@jjyao Could you please help me merge the PR? Last run failed with a flaky ray data test, which doesn't seem to be related to my change.

**Pull request merged**